Don't recommend 3 RTTs of credit (and other editorial) (#3301)
quicwg/base-drafts: draft-ietf-quic-transport.md

> ## Flow Credit Increments {#fc-credit}

Reviewer: How about "Managing Flow Control"?

Author: That seems too general. I'm not in love with what we have, but I prefer it because it's more specific.

Reviewer: I can see that.  These flow control sections are a bit too long for my liking, but it's good enough for now.

and MAX_DATA frames, but this section offers a few considerations.

To avoid blocking a sender, a receiver can send a MAX_STREAM_DATA or MAX_DATA
frame multiple times within a round trip or send it early enough to allow for
recovery from loss of the frame.

Control frames contribute to connection overhead. Therefore, frequently sending
MAX_STREAM_DATA and MAX_DATA frames with small changes is undesirable.  On the
other hand, if updates are less frequent, larger increments to limits are
necessary to avoid blocking a sender, requiring larger resource commitments at
the receiver.  There is a trade-off between resource commitment and overhead
when determining how large a limit is advertised.

A receiver can use an autotuning mechanism to tune the frequency and amount of
advertised additional credit based on a round-trip time estimate and the rate at
which the receiving application consumes data, similar to common TCP
implementations.  As an optimization, an endpoint could send frames related to
flow control only when there are other frames to send or when a peer is blocked,
ensuring that flow control does not cause extra packets to be sent.

> A sender that is blocked could choose to not send STREAM_DATA_BLOCKED or

Reviewer: ```suggestion
A blocked sender is not required to send STREAM_DATA_BLOCKED or
```

DATA_BLOCKED frames. Therefore, a receiver MUST NOT wait for a
STREAM_DATA_BLOCKED or DATA_BLOCKED frame before sending a MAX_STREAM_DATA or
MAX_DATA frame; doing so could result in the sender being blocked for the rest

> of the connection. Even if the sender sent these frames, waiting for them will

Reviewer: ```suggestion
of the connection. Even if the sender sends these frames, waiting for them will
```

result in the sender being blocked for at least an entire round trip.

> When a sender receives credit after being blocked, it can send a large amount of

Reviewer: ```suggestion
When a sender receives credit after being blocked, it may be able to send a large amount of
```

## Flow Credit Increments {#fc-credit}

Implementations decide when and how many bytes to advertise in MAX_STREAM_DATA
and MAX_DATA frames. This section describes one requirement and offers a few
considerations.

> A receiver MUST NOT wait for a STREAM_DATA_BLOCKED or DATA_BLOCKED frame before

Reviewer 2: I understand that we are moving existing text, but "MUST NOT" seems a bit odd to me, as it could be interpreted as a prohibition against the receiver intentionally blocking the stream. Maybe "SHOULD NOT" is better.

Author: Yes, it seems unnecessary. Let's do this separately in a design issue though, letting this one be editorial.

Reviewer 2: Ah that's a good point. Rereading the entire paragraph, I think I'm less concerned about readers possibly misinterpreting the text (though, admittedly, I was a bit confused).

Author: I've rephrased and moved the text around a bit, so hopefully it's clearer now.

Reviewer 2: Indeed it is. Thank you for the change.

> sending a MAX_STREAM_DATA or MAX_DATA frame, since doing so will mean that a

Reviewer: ```suggestion
sending a MAX_STREAM_DATA or MAX_DATA frame, since doing so means a sender will
```

> sender could be blocked for the rest of the connection if the peer chooses to

Reviewer: ```suggestion
be blocked for the rest of the connection if the peer chooses not to
```

> not send STREAM_DATA_BLOCKED or DATA_BLOCKED frames. Even if the peer sent these

Reviewer: ```suggestion
send STREAM_DATA_BLOCKED or DATA_BLOCKED frames. Even if the peer sent these
```

frames, waiting for them means that a sender will be blocked for at least an
entire round trip.

A sender that runs out of flow control credit will be unable to send new data
and is considered blocked, resulting in degraded performance for the

> connection. To avoid blocking a sender and to reasonably account for the

Reviewer: ```suggestion
connection. To avoid blocking a sender,
```

> possibility of loss, a receiver can send a MAX_STREAM_DATA or MAX_DATA frame

Reviewer: ```suggestion
a receiver can send a MAX_STREAM_DATA or MAX_DATA frame
```

multiple times within a round trip or send it early enough to allow for recovery

> from potential loss.

Reviewer: ```suggestion
from loss of the frame.
```

Control frames contribute to connection overhead. Therefore, frequently sending
MAX_STREAM_DATA and MAX_DATA frames with small changes is undesirable.  At the

> same time, larger increments to limits are necessary to avoid blocking if

Reviewer: I'd switch this around, ie:
"At the same time, less frequent updates require larger increments to the limits to avoid blocking."

updates are less frequent, requiring larger resource commitments at the
receiver.  There is a trade-off between resource commitment and overhead when
determining how large a limit is advertised.

A receiver can use an autotuning mechanism to tune the frequency and amount of
advertised additional credit based on a round-trip time estimate and the rate at
which the receiving application consumes data, similar to common TCP
implementations.  As an optimization, an endpoint could send frames related to
flow control only when there are other frames to send or when a peer is blocked,
ensuring that flow control does not cause extra packets to be sent.

When a sender receives credit after being blocked, it can send a large amount of

> data in response resulting in short-term congestion; see Section 6.9 in

Reviewer: ```suggestion
data in response, resulting in short-term congestion; see Section 6.9 in
```

flow control only when there are other frames to send or when a peer is blocked,
ensuring that flow control does not cause extra packets to be sent.

> 

Reviewer 3: I think this cross-reference is needed - people implementing flow control need to ensure this part is correctly implemented also.

Author: I've added a reference in but with better text I think.
